pkg/controller/velerobackup/velerobackup_controller_test.go

```diff
@@ -139,14 +161,168 @@ func TestReconcile(t *testing.T) {
+			assertHiveObjectsEqual(t, test.expectedHiveObjects, actualHiveObjects)
+			assert.Equal(t, test.expectedError, actualError)
+		})
+	}
+}
+
+func TestUpdateLastBackupChecksum(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+
+	tests := []struct {
+		name                string
+		hiveObjects         []*hiveObject
+		existingObjects     []runtime.Object
+		expectedHiveObjects []*hiveObject
+		expectedError       error
+	}{
+		{
+			name:                "No HiveObjects case",
+			hiveObjects:         []*hiveObject{},
+			existingObjects:     []runtime.Object{},
+			expectedHiveObjects: []*hiveObject{},
+			expectedError:       nil,
+		},
+		{
+			name: "1 changed ClusterDeployment case",
+			hiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			},
+			existingObjects: []runtime.Object{
+				testclusterdeployment.Build(changedClusterDeploymentBase()),
+			},
+			expectedHiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase(), testclusterdeployment.WithBackupChecksum(defaultChecksumFunc))),
+			},
+			expectedError: nil,
+		},
+		{
+			name: "1 changed SyncSet case",
+			hiveObjects: []*hiveObject{
+				ro2ho(testsyncset.Build(changedSyncSetBase())),
+			},
+			existingObjects: []runtime.Object{
+				testsyncset.Build(changedSyncSetBase()),
+			},
+			expectedHiveObjects: []*hiveObject{
+				ro2ho(testsyncset.Build(changedSyncSetBase(), testsyncset.WithBackupChecksum(defaultChecksumFunc))),
+			},
+			expectedError: nil,
+		},
+		{
+			name: "Multiple changed HiveObjects case",
+			hiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+				ro2ho(testsyncset.Build(changedSyncSetBase())),
+			},
+			existingObjects: []runtime.Object{
+				testclusterdeployment.Build(changedClusterDeploymentBase()),
+				testsyncset.Build(changedSyncSetBase()),
+			},
+			expectedHiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase(), testclusterdeployment.WithBackupChecksum(defaultChecksumFunc))),
+				ro2ho(testsyncset.Build(changedSyncSetBase(), testsyncset.WithBackupChecksum(defaultChecksumFunc))),
+			},
+			expectedError: nil,
+		},
+		{
+			name: "List error case",
+			hiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			},
+			existingObjects:     []runtime.Object{},
+			expectedError:       utilerrors.NewAggregate([]error{&errStatus}),
+			expectedHiveObjects: []*hiveObject{},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+			r := fakeClientReconcileBackup(test.existingObjects)
+
+			// Act
+			actualError := r.updateHiveObjectsLastBackupChecksum(test.hiveObjects)
+			actualHiveObjects, err := r.getHiveObjectsInNamespace(namespace, filterAllObjects)
+
+			// Assert
+			assert.NoError(t, err)
+			assertHiveObjectsEqual(t, test.expectedHiveObjects, actualHiveObjects)
```

Comment:
@staebler I'm fuzzy on what you're asking here.

Are you asking me to recreate the logic of `getHiveObjectsInNamespace` in the test code and return `runtime.Objects` instead of `hiveObjects`?

What's the problem with using non-test code to simplify / minimize test code?

I understand it's an implementation detail, but it's also a utility. I'm not understanding when it's ok to use a non-test code utility vs recreating the logic in test code.

Reply:
My issue is that the component being tested is `ReconcileBackup`. The `getHiveObjectsInNamespace` function is a big portion of what that component does. If you rely on that function to determine what the expected output is, then you are coupling the expected output to the behavior of the unit being tested. If there are errors in that function, it may result in tests that report success when they should fail.

Give me a few minutes, and I will post up a sample of what I am thinking would be a cleaner test.

Reply:
```
func TestUpdateLastBackupChecksum(t *testing.T) {
	apis.AddToScheme(scheme.Scheme)

	tests := []struct {
		name              string
		hiveObjects       []*hiveObject
		existingObjects   []runtime.Object
		expectedChecksums []string
		expectedError     error
	}{
		{
			name:              "No HiveObjects case",
			expectedChecksums: []string{},
		},
		{
			name: "1 changed ClusterDeployment case",
			hiveObjects: []*hiveObject{
				{
					object:   testclusterdeployment.Build(changedClusterDeploymentBase()),
					checksum: "new-checksum",
				},
			},
			existingObjects: []runtime.Object{
				testclusterdeployment.Build(changedClusterDeploymentBase()),
			},
			expectedChecksums: []string{
				"new-checksum",
			},
		},
		{
			name: "Multiple changed HiveObjects case",
			hiveObjects: []*hiveObject{
				{
					object:   testclusterdeployment.Build(changedClusterDeploymentBase()),
					checksum: "new-cd-checksum",
				},
				{
					object:   testsyncset.Build(changedSyncSetBase()),
					checksum: "new-syncset-checksum",
				},
			},
			existingObjects: []runtime.Object{
				testclusterdeployment.Build(changedClusterDeploymentBase()),
				testsyncset.Build(changedSyncSetBase()),
			},
			expectedChecksums: []string{
				"new-cd-checksum",
				"new-syncset-checksum",
			},
		},
		{
			name: "List error case",
			hiveObjects: []*hiveObject{
				{
					object:   testclusterdeployment.Build(changedClusterDeploymentBase()),
					checksum: "new-checksum",
				},
			},
			expectedChecksums: []string{},
			expectedError:     utilerrors.NewAggregate([]error{&errStatus}),
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			// Arrange
			r := fakeClientReconcileBackup(test.existingObjects)
			for _, ho := range test.hiveObjects {
				ho.logger = r.logger
			}

			// Act
			actualError := r.updateHiveObjectsLastBackupChecksum(test.hiveObjects)

			// Assert
			actualChecksums := make([]string, len(test.existingObjects))
			for i := range test.existingObjects {
				obj := test.existingObjects[i].DeepCopyObject()
				meta := obj.(metav1.Object)
				err := r.Get(context.TODO(), types.NamespacedName{Name: meta.GetName(), Namespace: meta.GetNamespace()}, obj)
				assert.NoError(t, err, "unexpected error fetching object")
				actualChecksums[i] = meta.GetAnnotations()[controllerutils.LastBackupAnnotation]
			}
			assert.Equal(t, test.expectedChecksums, actualChecksums, "unexpected checksums")
			assertErrorsAreEqualType(t, test.expectedError, actualError)
		})
	}
}
```

Reply:
The preceding test validates that the `updateHiveObjectsLastBackupChecksum` function updates the checksum of each of the runtime objects composed in the hive objects passed to the function.

Reply:
I would also accept the following.
1) Create a utility function separate from `ReconcileBackup`. This function should takes a client, a slice of type lists, and a predicate function to filter desired runtime objects and return a slice of runtime objects.
2) Create unit tests for the new utility function.
3) Have `getHiveObjectsInNamespace` call the new utility function.
4) Use the new utility function in the unit tests of `ReconcileBackup`.
However, even with that, the `ReconcileBackup` would be comparing runtime objects and not hive objects.

Reply:
Ok, yeah, I decided to do that. That feels cleaner to me. Please let me know if I implemented it how you were thinking.